c10/core/impl/cow/state_machine.h

```diff
@@ -0,0 +1,176 @@
+  ~StateMachine();
+
+  // Gets the current generation of the physical storage.
+  auto physical_generation() -> Generation;
+
+  // Bumps the generation if the shadow storage is non-null. If
+  // non-null, shadow storage must be the result of a previous call to
+  // simulate_copy_on_write on this instance.
+  //
+  // This will be called whenever we track a write to a Tensor. We
+  // will call this on the Tensor receiving the mutation, and it will
+  // propagate through its storage to here. That tensor will send its
+  // shadow storage here, but it will be null if it was the original
+  // tensor or a view derived from it.
+  auto maybe_bump(cow::ShadowStorage* maybe_shadow_storage) -> void;
+
+  // Simulates a lazy copy a tensor that owns the shadow storage.
+  //
+  // maybe_shadow_storage comes from the tensor that is being lazily
+  // copied. This may be null if this is the first lazy copy taking
+  // place, or if the lazy copy is being performed on a tensor that
+  // was part of the original tensors that share a view.
+  //
+  // The generation of the output will come from:
+  // 1) maybe_shadow_storage->generation(), if non-null.
+  // 2) this->default_shadow_storage_->generation(), if it is set.
+  // 3) physical_generation_, i.e. 0, if this is the first lazy copy.
+  auto simulate_lazy_copy(cow::ShadowStorage* maybe_shadow_storage)
+      -> intrusive_ptr<cow::ShadowStorage>;
+
+ private:
+  class Impl;
+
+  /** Gets the underlying Impl, returning null if uninitialized. */
+  auto maybe_get_impl() -> Impl*;
+  /** Gets the underlying Impl, initializing if uninitialized. */
+  auto ensure_initialized() -> Impl&;
+
+  /// Gets the Impl, invoking the callback if uninitialized.
+  ///
+  /// on_initial_func should return c10::optional<Impl*>. Return:
+  ///  * nullopt if we should continue in the loop, e.g. waiting for
+  ///    another thread to initialize
+  ///  * nullptr if we should return no impl to the user
+  ///  * a pointer to an implementation, if it exists
+  template <typename OnInitialFunc>
+  auto get_impl(const OnInitialFunc& on_initial_func) -> Impl*;
+
+  // Identifies where a storage instance is in the lifecycle of
+  // tracking simulated lazy copies.
+  enum class StateId {
+    // This is the initial state. It may not be returned to and
+    // indicates that the storage has never had a lazy copy requested
+    // of it.
+    //
+    // INVARIANT: maybe_get_impl() == nullptr
+    initial = 2,
+
+    // A transition state between initial and active. Any visitor that
+    // sees this state will spin until it sees the active state.
+    initializing = 1,
+
+    // This is transitioned to when the first copy on write takes
+    // place.
+    //
+    // Subtle! This has the value of 0 because we are using the
+    // alignment bits of a pointer to hold this state id. In the case
+    // where we have an actual impl, the state id will be active,
+    // hence the alignment bits will be 0.
+    active = 0,
```

Comment:
nit: you probably want some sort of assert that says ensures this enum doesn't get too big, otherwise you'll run out of bits.

Reply:
moot!